Mixpanel/Mixpanel.m

```diff
@@ -1852,6 +1856,107 @@ - (BOOL)enableLogging
     }
 }
 
+#pragma mark - Mixpanel Push Notifications
+
++ (BOOL)isMixpanelPushNotification:(UNNotificationContent *)content {
+    if ([content userInfo] == nil) {
+        MPLogInfo(@"%@ userInfo was nil, returning false");
+        return false;
+    }
+    return [content.userInfo objectForKey:@"mp"] != nil;
+}
+
++ (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler {
+
+    if (![self isMixpanelPushNotification:response.notification.request.content]) {
+        MPLogWarning(@"%@Calling MixpanelPushNotifications.handleResponse on a non-Mixpanel push notification is a noop", self);
+        completionHandler();
+        return;
+    }
+
+    NSDictionary *userInfo = response.notification.request.content.userInfo;
+
+    // Initialize properties to track to Mixpanel
+    NSMutableDictionary *trackingProps = [[NSMutableDictionary alloc] init];
+    [trackingProps setValuesForKeysWithDictionary:@{
+        @"campaign_id": [userInfo valueForKeyPath:@"mp.c"],
+        @"message_id": [userInfo valueForKeyPath:@"mp.m"],
+    }];
+
+
+    MPLogInfo(@"%@ didReceiveNotificationResponse action: %@", self, response.actionIdentifier);
+
+    // If the notification was dismissed, just track and return
+    if ([response.actionIdentifier isEqualToString:UNNotificationDismissActionIdentifier]) {
+        [instances.allKeys enumerateObjectsUsingBlock:^(NSString *token, NSUInteger idx, BOOL * _Nonnull stop) {
+            Mixpanel *instance = instances[token];
+            [instance track:@"$push_notification_dismissed" properties:trackingProps];
+            [instance flush];
+        }];
+        completionHandler();
+        return;
+    }
+
+    NSDictionary *ontap = nil;
+
+    if ([response.actionIdentifier isEqualToString:UNNotificationDefaultActionIdentifier]) {
+        // The action that indicates the user opened the app from the notification interface.
+        [trackingProps setValue:@"notification" forKey:@"tap_target"];
+        if (userInfo[@"mp_ontap"]) {
+            ontap = userInfo[@"mp_ontap"];
+        }
+    } else {
+        // Non-default, non-dismiss action -- probably a button tap
+        BOOL wasButtonTapped = [response.actionIdentifier containsString:@"MP_ACTION_"];
+        if (wasButtonTapped) {
+            NSArray *buttons = userInfo[@"mp_buttons"];
+            NSInteger idx = [[response.actionIdentifier stringByReplacingOccurrencesOfString:@"MP_ACTION_" withString:@""] integerValue];
+            NSDictionary *buttonDict = buttons[idx];
+            ontap = buttonDict[@"ontap"];
+            [trackingProps setValuesForKeysWithDictionary:@{
+                @"button_id": buttonDict[@"id"],
```

Comment:
might worth adding safe guard check for nil for buttonDict[@"id"] and buttonDict[@"lbl"], it will crash if set nil to dictionary.